### PR TITLE
[CUDA] 64-bit indexing for avg_pool_backward

### LINKED ISSUE #113833: cannot backprop a cnn when intermediate output has size larger than 2**31
### 🐛 Describe the bug

This might be related to a few previous issues like #43476, #51871. I find that when my modified ResNet has the size of any of its intermediate layers to be larger than 2**31 (max value of int32), which can happen if you have 100 batch size, 200 channels, 512*512 images, it may raise an error when you try to backprop your loss.

See below for a minimal example to reproduce the issue, although you may want to run it on a large memory GPU, otherwise you'll first get the GPU out of memory error before this int32 error.

I checked that with 80GB A100 the GPU memory is not the real issue; the restriction of int32 unnecessarily limits the application of CNN to high resolution images. Please consider use int64 instead; I suppose it just needs a switch of data type that will not deteriorate the performance.

```python
import numpy as np
import torch
import torch.nn as nn

x = torch.rand(100, 50, 512, 512).cuda()
c = nn.Conv2d(50, 200, kernel_size=1).cuda()
r = nn.ReLU().cuda()

loss = torch.sum(r(c(x)))
print(loss)
loss.backward() # this is fine though

loss = torch.prod(r(c(x)))
print(loss)
loss.backward() # this line gives you the error
```

The last line leads to the following error,

```
>>> loss.backward()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hejia/.conda/envs/hejia@della-5/lib/python3.11/site-packages/torch/_tensor.py", line 492, in backward
    torch.autograd.backward(
  File "/home/hejia/.conda/envs/hejia@della-5/lib/python3.11/site-packages/torch/autograd/__init__.py", line 251, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: nonzero is not supported for tensors with more than INT_MAX elements,   file a support request
```

FYI, below is the actual error message I got from a real world example, which I believe has the same origin as the minimal toy example above.

```
Traceback (most recent call last):
  File "/home/hejia/lss-dev/00-wl/04/01-train-0-10k.py", line 266, in <module>
    foo = nqe.train_1d(qnet[i_dim], device='cuda', train_loader=train_loader, valid_loader=valid_loader,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hejia/nqe/nqe/train.py", line 283, in train_1d
    loss_now.backward()
  File "/home/hejia/.conda/envs/hejia@della-5/lib/python3.11/site-packages/torch/_tensor.py", line 492, in backward
    torch.autograd.backward(
  File "/home/hejia/.conda/envs/hejia@della-5/lib/python3.11/site-packages/torch/autograd/__init__.py", line 251, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: integer out of range
```

### Versions

(hejia@della-5) [hejia@della-l07g3 ~]$ python collect_env.py 
Collecting environment information...
PyTorch version: 2.1.1
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Springdale Linux release 8.8 (Modena) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18)
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.28

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-477.27.1.el8_8.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 535.104.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
OS: Springdale Linux release 8.8 (Modena) (x86_64)
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              48
On-line CPU(s) list: 0-47
Thread(s) per core:  1
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          25
Model:               1
Model name:          AMD EPYC 7413 24-Core Processor
Stepping:            1
CPU MHz:             2650.000
CPU max MHz:         3630.8101
CPU min MHz:         1500.0000
BogoMIPS:            5290.31
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-23
NUMA node1 CPU(s):   24-47
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpui
d extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw 
ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx 
smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean 
flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.1.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.16.1
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.8           py311h5eee18b_0  
[conda] mkl_random                1.2.4           py311hdb19cb5_0  
[conda] numpy                     1.26.0          py311h08b1b3b_0  
[conda] numpy-base                1.26.0          py311hf175353_0  
[conda] pytorch                   2.1.1           py3.11_cuda12.1_cudnn8.9.2_0    pytorch
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchtriton               2.1.0                     py311    pytorch
[conda] torchvision               0.16.1              py311_cu121    pytorch

cc @ptrblck

### PR DESCRIPTION
**Pull request by eqy**
Fixes #113833

cc @ptrblck @mikaylagawarecki

### CODE REVIEW

**Comment by eqy** [comment]
CC @malfet

**Review by malfet — APPROVED** [review]
LGTM

File: aten/src/ATen/native/cuda/AveragePool2d.cu [hunk 1]
```diff
@@ -183,8 +183,8 @@ __global__ void avg_pool2d_backward_out_cuda_frame(const int nthreads, const sca
   }
 }
 
-template <typename scalar_t, typename accscalar_t>
-__global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const int nthreads,
+template <typename scalar_t, typename accscalar_t, typename index_t>
+__global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const index_t nthreads,
     const scalar_t* const top_diff,
     const int64_t channels, const int64_t height,
     const int64_t width, const int pooled_height, const int pooled_width,
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/cuda/AveragePool2d.cu:190]: noob question that is maybe unrelated: why are `pooled_height` and `pooled_width` `const int` here in `avg_pool2d_backward_out_cuda_frame_nhwc` but `const int64_t` above in  `avg_pool2d_backward_out_cuda_frame`?

Comment (eqy) [reply aten/src/ATen/native/cuda/AveragePool2d.cu:190]: It doesn't look obvious to me as I can't immediately see where its use/definition diverges between the channels-first and channels-last cases. Maybe @xwang233 can comment as he authored the channels-last version (https://github.com/pytorch/pytorch/pull/35855) and the last round of 64-bit indexing fixes.

Comment (mikaylagawarecki) [reply aten/src/ATen/native/cuda/AveragePool2d.cu:190]: btw I think my question is unrelated to the initial issue, did not mean to block this PR from merging. Could you perhaps file an issue about the typing difference between channels-last and non-channels-last for avg_pool though so this doesn't get lost please?

**Pull request closed without merging**